### ISSUE TITLE
Has secretsdump ever crashed a domain controller?

### ISSUE THREAD

**Issue by 7MinSec** [fortra/impacket#1436]
### Configuration  
impacket version:  Latest
Python version:  3.10
Target OS:  Kali latest

### Debug Output With Command String  
I was doing a secretsdump with DA creds as follows:

`./secretsdump.py -just-dc-ntlm DOMAIN/USER@IP.OF.DOMAIN.CONTROLLER -outputfile /path/to/some/file/name`

The dump started scrolling by as normal for about 30 seconds, then stopped on one particular line and just hung for about 30 seconds, followed by:

````
[-] timed out
[*] Something went wrong with the DRSUAPI approach. Try again with -use-vss parameter
[*] Cleaning up...
````

I used CME just to quickly revalidate the DC was up and accepting my creds, and I got dozens of lines of an exception/error.  I talked to the client and they found the DC had rebooted.  I did not run secretsdump again after that since I had a small heart attack.

Looking through the issues here and doing some Googling, I don't see other instances of this happening.  Any thoughts on what might've caused it, or what info you might want to know about the target DC?  All I know at this point is it is a Win Server 2016 Standard.

Thanks,
Brian

### PCAP  
Sorry, don't have one.

### Additional context  
None at this time.

**Comment by jagotu** [comment]
@7MinSec The only time I've seen this happen is with an EDR. Apparently there was a "feature" where if it detected credential dumping the EDR killed the whole DC to stop it. So if you know of any security products they use on the DC I'd start there.

**Comment by 7MinSec** [comment]
@jagotu thank you for this insight.  I'm looking to get the EDR logs now to see if it looks like that played a part.  I did get a peek at the event logs, and I believe this one is super telling:

````
EventID 1074:
The process wininit.exe has initiated the restart of computer XXX on behalf of user for the following reason: No title for this reason could be found

 Reason Code: 0x50006

 Shutdown Type: restart

 Comment: The system process 'C:\Windows\system32\lsass.exe' terminated unexpectedly with status code -1073741819.  The system will now shut down and restart.
````
In doing some Google-Fu, most of the buzz around some of these search terms point back to a faulty Jan 2022 update (http://www.edugeek.net/forums/windows/225791-looks-like-january-windows-updates-causing-all-sorts-issues.html).  

I'll report back when I know about the EDR situation.

**Comment by 7MinSec** [comment]
Update: no EDR installed on the DC.

**Comment by jagotu** [comment]
Interesting. I've had the exact same issue as you (lsass crash to reboot after ~30 seconds of successful dumping) once and thought it was EDR, but seems maybe it actually wasn't 🤔

I'll try to look up my notes but probably will not be able to add more than my anecdotal "I've seen it happen, too" report.

**Comment by BusyR** [comment]
I have seen the exact same issue, same events in the eventlog, on different 2 occasions. Not directly running secretsdump, but using "crackmapexec smb dc01 -u user -p pass --ntds". Both DC's ran Win2012R2.

**Comment by DeserranoJorden** [comment]
I'm having the same issue here.

I idid however had different DCs where I could test this on.
Note that I tested both "cme --ntds" and secretsdump.py.

The DCs that the full dump was possible on were normal Server 2012 R2 DCs (normal setup).
The DCs (tested on 2) that were rebooting with the same error messages as above were both Server 2019 and were hosted on Azure. 

I was unable to determine the reason of the reboot.
No real smoking gun on the system logs as well.

**Comment by professor-moody** [comment]
Having the same issue here.  Haven't noticed it until recently but it may have been happening for a while.  It happened on both server 2019 and 2016, in two separate environments.  Neither were in Azure as far as I know.  I suspect EDR but when attempting to lab out the environments, could not replicate...

**Comment by twosevenzero** [comment]
@DeserranoJorden @BusyR @7MinSec In your cases, was the domain controller also running Rapid7's InsightIDR agent `ir_agent.exe`?

**Comment by BusyR** [comment]
No, no Rapid7 EDR running on the DC's, but both DC's were running 
Symantec AV.

**Comment by twosevenzero** [comment]
> No, no Rapid7 EDR running on the DC's, but both DC's were running Symantec AV.

I have verified that it happens on a fresh install of Windows and DC Promotion.

**Comment by Gunbird** [comment]
Confirm here also on a Server 2019.

It happens only after a while so if the directory does not contain enough accounts it may not happen.

In my testing It could go as far as 19K accounts before rebooting.

I have also experienced the reboot with windows server 2012 R2 even though it seems to trigger later compared to 2019.

**Comment by twosevenzero** [comment]
Can we get this issue tagged as a bug/request instead of a question, please?

After additional testing, I am not as convinced as I was in the beginning. Now, I am pretty sure this IS an issue with the way secretsdump performs the dcsync.

Using other tools like dsinternals and mimikatz to do full syncs do not result in a crash of the domain controller. Examining the logs on the domain controller also show that there is a login attempt for each and every user while using secretsdump. This is obviously going to put a bunch of stress on LSASS. Here is a copy of the Windows Logs/Security Event 4662 log.

```
An operation was performed on an object.

Subject :
	Security ID:		CRASH\Administrator
	Account Name:		Administrator
	Account Domain:		CRASH
	Logon ID:		0xA6753C

Object:
	Object Server:		DS
	Object Type:		domainDNS
	Object Name:		DC=crash,DC=lab
	Handle ID:		0x0

Operation:
	Operation Type:		Object Access
	Accesses:		Control Access
				
	Access Mask:		0x100
	Properties:		Control Access
		{1131f6ad-9c07-11d1-f79f-00c04fc2dcd2}
	domainDNS


Additional Information:
	Parameter 1:		-
	Parameter 2:		
```

These happen about 80-100 times per second while doing a full dcsync as described in the logs I included in my previous post.

**Who at Fortra is in charge of this project now? Can I get someone to take a look at the code to see if there is a way to reduce these connections?**

### Hey there, Pentesters!
For the pentesters that are finding this after getting crushed by a client for bringing down a domain controller, I would switch to using mimikatz or the dsinternals powershell module to get the same result.

To do this you will need a runas session for the account you want to use but it works as expected once you are set up.

#### Installing DSInternals
```
Install-Module DSInternals -Force
```

Once this is installed you can use the `Get-ADReplAccount` to perform your DCSync. Obviously, you will need to read into how this works and add flags based on your environment. An example to get started with would be something like:

```
Get-ADReplAccount -Server dc01.crash.lab -All | Format-Custom -View PwDump
```

**Comment by mpgn** [comment]
I've tested on a brand new ad server 2022. Added 20k user and try to dump using secretdump.py the NTDS. No crash, no error. (defender was enabled).

The windows server was up to date.

**Comment by twosevenzero** [comment]
@mpgn After multiple passes with secretsdump against a fresh 2022 domain controller, I am seeing the same things you are. No crashes. 👍 However, this is still an issue for Windows Server 2012r2, 2016, and 2019 domain controllers which are significantly more prevalent in live corporate environments where many penetration testers are using the tools. Have you been able to test these to see if you are able to replicate the issue?

Thanks a ton for your input here. As a core contributor and developer of security tools, I am sure your voice is regarded much higher than mine and you are likely able to see what the issues may be faster than I will since you are intimately familiar with it all. Looking forward to your test results. ❤️

**Comment by mpgn** [comment]
I was able to reproduce on a windows server 2019, the error is not always the same but the DC crash yep.
The domain controller was brand new and up to date.

![image](https://user-images.githubusercontent.com/5891788/226472988-bad03e6b-131b-4bc1-add9-3f50c155476b.png)
![image](https://user-images.githubusercontent.com/5891788/226473007-8e1504b5-d4e3-4d95-95db-200d4a634e48.png)

**Comment by twosevenzero** [comment]
@snovvcrash @0xdeaddood @gabrielg5  Can we get this marked as a bug instead of a question? I would create a new issue but there is so much valuable data in this ticket.

**Comment by CCob** [comment]
Just my 2 pence here.  If you can crash lsass remotely via DRS, this is a CVE.  INVALID_POINTER_READ suggests a dangling pointer, so could potentially be converted to RCE

**Comment by twosevenzero** [comment]
Microsoft closed my ticket about LSASS.
I tested it today after the 04-2023 patches were applied and it still crashes so I guess they do not care.
Feel free to weaponize it @CCob

**Comment by anadrianmanrique** [comment]
It seems the issue was known for at least 5 years https://twitter.com/SkelSec/status/1042560510061932546 https://github.com/skelsec/windows_ad_dos_poc. 
Buggy function "draXlateNativeReplyToOutboundReply" was ported to windows 2019 
![image](https://user-images.githubusercontent.com/60896207/231623050-ab11fda0-9e63-4a7d-876a-385e7d14e582.png)

There are some slight modifications in windows 2022 though:
![image](https://user-images.githubusercontent.com/60896207/231623233-4cf38ee4-6a23-4897-a81a-e83dcfdaaf16.png)

**Comment by Danielg75** [comment]
Can we use secretsdump on assignments safely? or is it still a risk of BSOD on servers.

**Comment by professor-moody** [comment]
Revisited this issue today with a few colleagues and after _extensive testing_ confirmed 2016 and 2019 are still crashing.  More testing needs to be done on 2012 but presumably it's also still affected.

**Comment by twosevenzero** [comment]
> Can we use secretsdump on assignments safely? or is it still a risk of BSOD on servers.

I am sure you just meant crash, but it's important to know that the DC does not BSOD. The LSASS process crashes and the OS catches that. It then displays an error stating that the system is unstable and will reboot in 30 or 60 seconds. Reboots in most modern environments take like 20 seconds so you will only see this much downtime making it very easy to think there was just some kind of network hiccup. 
I would bet that teams across the world are just crashing LSASS and rebooting DCs all the time.

**Comment by anadrianmanrique** [comment]
Ok let me dump in an organized way, what I've been observing from my testing these days:

- **_secretsdump_** it's prone to crash lsass on environments with thousands of user accounts (tested against an up to date 2019). It never crashed windows 2022 though.
- Microsoft has been **reported** regarding this issue, but patch has never been released.
- impacket's **dcsync** implementation potentially, causes stress of the remote **lsass** process, as 1 rpc call (actually 2 calls) is issued per user account. This implementation seems to be replicated in several open source projects on github, and even ported to other languages (aiosmb, metasploit ,etc).
- Mimikatz's implementation, on the other hand, issues 1 rpc call every 1000 objects, which at first, looks like a more straightforward way to handle thousands of accounts https://github.com/stjordanis/mimikatz/blob/6910c7b930b9f1115d2fe1623c906f51e8e85311/mimikatz/modules/lsadump/kuhl_m_lsadump_dc.c#L69
- I've been able to crash the remote lsass with mimikatz ( consider that mimikatz uses the microsoft stack in order to request/response and serialize rpc )

**Comment by anadrianmanrique** [comment]
I managed to port mimikat'z implementation to impacket. I'm currently testing a PoC in order to eventually submit a PR. So far I encountered, in general, an improvement in performance and stability of the remote lsass process, BUT:

- the approach of requesting 1000 objects per call is not possible yet (at least in my lab) as there's a bug in impacket NDR deserialization process. This causes a crash in secretsdump. By adjusting the amount of objects to request, I was able to avoid the crash and perform the dcsync request successfully.
- In the middle of the development of the PoC I've got lsass crashed, even with the new approach.

**Comment by CN-Syndra** [comment]
> Ok let me dump in an organized way, what I've been observing from my testing these days:
> 
> * **_secretsdump_** it's prone to crash lsass on environments with thousands of user accounts (tested against an up to date 2019). It never crashed windows 2022 though.
> * Microsoft has been **reported** regarding this issue, but patch has never been released.
> * impacket's **dcsync** implementation potentially, causes stress of the remote **lsass** process, as 1 rpc call (actually 2 calls) is issued per user account. This implementation seems to be replicated in several open source projects on github, and even ported to other languages (aiosmb, metasploit ,etc).
> * Mimikatz's implementation, on the other hand, issues 1 rpc call every 1000 objects, which at first, looks like a more straightforward way to handle thousands of accounts https://github.com/stjordanis/mimikatz/blob/6910c7b930b9f1115d2fe1623c906f51e8e85311/mimikatz/modules/lsadump/kuhl_m_lsadump_dc.c#L69
> * I've been able to crash the remote lsass with mimikatz ( consider that mimikatz uses the microsoft stack in order to request/response and serialize rpc )

When I used mimikatz, in a domain with 50,000+ user, it caused lsass to crash and the DC to restart.   mimikatz doesn't seem to solve the problem

**Comment by anadrianmanrique** [comment]
fixed in #2017. please reopen if needed
